### PR TITLE
[FLINK-2066][core] Add delay between execution retires configuration …

### PR DESCRIPTION
**Pull request by WangCHX**
…at job level

want to solve FLINK-2066. Based on when there has number of retries, there must has retry delay. Add field to ExecutionConfig.java. Then use eclipse find places which call numberOfRetries, at these places, I add retryDelay.

I am very new to distributed system. And this is the first time I contribute on open source. sorry for any inconvenience.

### CODE REVIEW

**Comment by rmetzger** [comment]
Thanks a lot for your contribution to Flink!
Sorry for not giving you feedback earlier. Many committers are currently busy preparing the next Flink release.

Can you also update the documentation about the Execution Conf ? There is an entire section on the available parameters:
https://ci.apache.org/projects/flink/flink-docs-master/apis/programming_guide.html#execution-configuration
You can find the markdown files for the documentation in the `docs/` directory of the Flink source.

I quickly looked over the code, and I think its in a good shape!

**Comment by WangCHX** [comment]
Thank you. Sure. I will do it.

**Comment by WangCHX** [comment]
Thank you very much.

File: docs/apis/programming_guide.md [hunk 1]
```diff
@@ -1992,6 +1992,8 @@ With the closure cleaner disabled, it might happen that an anonymous user functi
 
 - `getNumberOfExecutionRetries()` / `setNumberOfExecutionRetries(int numberOfExecutionRetries)` Sets the number of times that failed tasks are re-executed. A value of zero effectively disables fault tolerance. A value of `-1` indicates that the system default value (as defined in the configuration) should be used.
 
+- `getExecutionRetryDelay()` / `setExecutionRetryDelay(long executionRetryDelay)` Sets the delay that failed tasks are re-executed. A value of `-1` indicates that the default value should be used.
```
Comment (StephanEwen) [inline docs/apis/programming_guide.md]: I think this is a critical parameter, so I would like to extend the description a bit. How about this:

```
Sets the delay that the system waits after a job has failed, before re-executing it. The delay starts after all tasks have been successfully been stopped on the TaskManagers, and once the delay is past, the tasks are re-started. This parameter is useful to delay re-execution in order to let certain time-out related failures surface fully (like broken connections that have not fully timed out), before attempting a re-execution and immediately failing again due to the same problem.  

This parameter only has an effect if the number of execution re-tries is one or more.
```

File: flink-core/src/main/java/org/apache/flink/api/common/Plan.java [hunk 2]
```diff
@@ -303,6 +303,13 @@ public int getNumberOfExecutionRetries() {
 	}
 	
 	/**
+	 * Gets the delay between retry failed task.
+	 * @return The delay the system will wait to retry.
+	 */
+	public long getExecutionRetryDelay(){
```
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/Plan.java]: Style: Insert a space between `()` and `{`

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 3]
```diff
@@ -199,6 +199,31 @@ public int getNumberOfExecutionRetries() {
...
+	 * Sets the delay that failed tasks are re-executed. A value of
+	 * zero effectively disables fault tolerance. A value of {@code -1}
+	 * indicates that the system default value (as defined in the configuration)
+	 * should be used.
+	 *
+	 * @param executionRetryDelay
+	 * 		The delay of time the system will wait to re-execute failed
+	 * 		tasks.
+	 */
+	public void setExecutionRetryDelay(long executionRetryDelay){
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: See above

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 4]
```diff
@@ -199,6 +199,31 @@ public int getNumberOfExecutionRetries() {
...
+	}
+	
+	/**
+	 * Gets the delay time in milliseconds the system will wait to re-execute failed tasks.
+	 * A value of {@code -1} indicates that the system default value (as defined
+	 * in the configuration) should be used.
+	 *
+	 * @return The delay time the system will wait to re-execute failed tasks.
+	 */
+	public long getExecutionRetryDelay(){
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: See above

**Comment by StephanEwen** [comment]
A good start.

I think for the batch API, there is a step missing, which sets the parameter on the `JobGraph`, in the `JobGraphGenerator`. The streaming and batch program translation parts are a bit different, because the batch code path applies additionally the Optimizer.

File: docs/apis/programming_guide.md (hunk 1 above)
Comment (WangCHX) [inline docs/apis/programming_guide.md]: much better than mine. Thanks.

File: flink-core/src/main/java/org/apache/flink/api/common/Plan.java (hunk 2 above)
Comment (WangCHX) [inline flink-core/src/main/java/org/apache/flink/api/common/Plan.java]: fixed.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java (hunk 3 above)
Comment (WangCHX) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: fixed.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java (hunk 4 above)
Comment (WangCHX) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: fixed.

**Comment by StephanEwen** [comment]
Thanks for addressing the comments!

Looks very good now, +1 to merge this

**Comment by fhueske** [comment]
OK, will merge this one. Thanks for the quick update @WangCHX!

**Pull request closed without merging**